x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/GetDeploymentStatsAction.java

```diff
@@ -0,0 +1,237 @@
+
+    private GetDeploymentStatsAction() {
+        super(NAME, GetDeploymentStatsAction.Response::new);
+    }
+
+    public static class Request extends BaseTasksRequest<GetDeploymentStatsAction.Request> {
+
+        public static final String ALLOW_NO_MATCH = "allow_no_match";
+
+        private final String deploymentId;
+        private boolean allowNoMatch = true;
+        // used internally this should not be set by the REST request
+        private List<String> expandedIds;
+
+        public Request(String deploymentId) {
+            this.deploymentId = ExceptionsHelper.requireNonNull(deploymentId, "deployment_id");
+            this.expandedIds = Collections.singletonList(deploymentId);
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            this.deploymentId = in.readString();
+            this.allowNoMatch = in.readBoolean();
+            this.expandedIds = in.readStringList();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(deploymentId);
+            out.writeBoolean(allowNoMatch);
+            out.writeStringCollection(expandedIds);
+        }
+
+        public String getDeploymentId() {
+            return deploymentId;
+        }
+
+        public List<String> getExpandedIds() {
+            return expandedIds;
+        }
+
+        public void setExpandedIds(List<String> expandedIds) {
+            this.expandedIds = expandedIds;
+        }
+
+        public void setAllowNoMatch(boolean allowNoMatch) {
+            this.allowNoMatch = allowNoMatch;
+        }
+
+        public boolean isAllowNoMatch() {
+            return allowNoMatch;
+        }
+
+            @Override
+        public boolean match(Task task) {
+            return expandedIds.stream().anyMatch(taskId -> StartTrainedModelDeploymentAction.TaskMatcher.match(task, taskId));
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Request request = (Request) o;
+            return Objects.equals(deploymentId, request.deploymentId) &&
+                this.allowNoMatch == request.allowNoMatch &&
+                Objects.equals(expandedIds, request.expandedIds);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(deploymentId, allowNoMatch, expandedIds);
+        }
+    }
+
+
+    public static class Response extends BaseTasksResponse implements ToXContentObject {
+
+        public static final ParseField DEPLOYMENT_STATS = new ParseField("deployment_stats");
+
+        public static class DeploymentStats implements ToXContentObject, Writeable {
+
+            private final String modelId;
+            private final String node_id;
```

Comment:
We should probably remove the `node_id` as it will not be possible to have when we end up allocating the same model on multiple nodes.

Reply:
Our other stats APIs return a whole `node` object (search for "Properties of node" in https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-get-job-stats.html to see what this contains).  `node_id` is not friendly, as it's the auto-generated unique ID, not the friendly node name.  I think we should follow the pattern of our other stats APIs and return `node` objects.  But in this case it needs to be a `nodes` array so it works both in the current persistent task world and the future model allocation world.

Reply:
I see this becoming a map of stats under model & node  e.g.

```
model_id: {
    model_size: 1.5 kb,
    nodes: [ {
           node_id: XX,
           average_inference_time_ms: 3.0,
	   last_access_string: 2021-07-12 T22: 13: 31.022 Z,
	   inference_count: 1
        }, {
           node_id: YY,
           average_inference_time_ms: 4.0,
           ....
   ]
}
```

The final output will be shaped by the work on model allocation and is likely to be very different to what is here

Reply:
I've pushed a commit to add the node object and restructured the response as above